### PR TITLE
[IMP] event sale newapi 

### PR DESCRIPTION
**Pull request by DudhatShivam**
Task: https://www.odoo.com/web#id=12010&view_type=form&model=project.task&menu_id=3942&action=327

Pad: https://pad.odoo.com/p/r.Hddf0qvcf9AJY1fL

### CODE REVIEW

File: addons/event_sale/models/sale_order.py [hunk 1]
```diff
@@ -1,53 +1,45 @@
...
-            event_ctx = dict(context, default_sale_order_id=ids[0])
-            return self.pool.get('ir.actions.act_window').for_xml_id(cr, uid, 'event_sale', 'action_sale_order_event_registration', event_ctx)
-        else:
-            return res
+    @api.multi
+    def action_button_confirm(self):
+        self.ensure_one()
+        super(SaleOrder, self).action_button_confirm()
+        if self.mapped('order_line.event_id'):
+            return self.env['ir.actions.act_window'].with_context(default_sale_order_id=self.id).for_xml_id('event_sale', 'action_sale_order_event_registration')
```
Comment (xmo-odoo) [inline addons/event_sale/models/sale_order.py]: Is it normal that the result of the `super()` call is ignored and nothing is returned if tehere are no `order_line.event_id`?

File: addons/event_sale/data/event_demo.xml [hunk 2]
```diff
@@ -1,7 +1,5 @@
 <?xml version="1.0"?>
```
Comment (rgo-odoo) [inline addons/event_sale/data/event_demo.xml]: you missed encoding

File: addons/event_sale/models/event.py [hunk 3]
```diff
@@ -25,11 +20,21 @@ def _default_tickets(self):
         except ValueError:
             return self.env['event.event.ticket']
```
Comment (rgo-odoo) [inline addons/event_sale/models/event.py]: use raise_if_not_found=False in ref(...) and change the code accordingly.

File: addons/event_sale/models/event.py [hunk 4]
```diff
@@ -25,11 +20,21 @@ def _default_tickets(self):
...
 
+    event_ticket_ids = fields.One2many('event.event.ticket', 'event_id', string='Event Ticket', default=lambda rec: rec._default_tickets(), copy=True)
+
 
-class event_ticket(models.Model):
+class EventTicket(models.Model):
     _name = 'event.event.ticket'
     _description = 'Event Ticket'
 
+    @api.model
```
Comment (rgo-odoo) [inline addons/event_sale/models/event.py]: decorator not required.

File: addons/event_sale/models/event.py [hunk 5]
```diff
@@ -25,11 +20,21 @@ def _default_tickets(self):
...
+    def _default_product_id(self):
+        try:
+            product = self.env['ir.model.data'].get_object('event_sale', 'product_product_event')
```
Comment (rgo-odoo) [inline addons/event_sale/models/event.py]: use self.env.ref(...)

File: addons/event_sale/models/product.py [hunk 6]
```diff
@@ -1,29 +1,28 @@
...
-    def onchange_event_ok(self, cr, uid, ids, type, event_ok, context=None):
-        if event_ok:
-            return {'value': {'type': 'service'}}
-        return {}
+    event_ok = fields.Boolean(string='Event Subscription', help='Determine if a product needs to create automatically an event registration at the confirmation of a sales order line.')
+    event_type_id = fields.Many2one('event.type', string='Type of Event',
+                                    help='Select event types so when we use this product in sales order lines, it will filter events of this type only.')
 
+    @api.onchange('event_ok')
+    def onchange_event_ok(self):
```
Comment (rgo-odoo) [inline addons/event_sale/models/product.py]: onchange_event_ok --> _onchange_event_ok

File: addons/event_sale/models/sale_order.py [hunk 7]
```diff
@@ -1,44 +1,32 @@
...
-        for order in self.browse(cr, uid, ids, context=context):
-            redirect_to_event_registration, so_id = any(line.event_id for line in order.order_line), order.id
-            order.order_line._update_registrations(confirm=True)
-        if redirect_to_event_registration:
-            event_ctx = dict(context, default_sale_order_id=so_id)
-            return self.pool.get('ir.actions.act_window').for_xml_id(cr, uid, 'event_sale', 'action_sale_order_event_registration', event_ctx)
-        else:
-            return res
+    @api.multi
+    def action_confirm(self):
```
Comment (rgo-odoo) [inline addons/event_sale/models/sale_order.py]: add self.ensure_one()

File: addons/event_sale/models/sale_order.py [hunk 8]
```diff
@@ -1,44 +1,32 @@
...
+        res = super(SaleOrder, self).action_confirm()
```
Comment (rgo-odoo) [inline addons/event_sale/models/sale_order.py]: why you removed the call to _update_registrations() ?

File: addons/event_sale/models/sale_order.py [hunk 9]
```diff
@@ -1,44 +1,32 @@
...
-            help="Choose an event ticket and it will automatically create a registration for this event ticket."),
-        # those 2 fields are used for dynamic domains and filled by onchange
-        # TDE: really necessary ? ...
-        'event_type_id': fields.related('product_id', 'event_type_id', type='many2one', relation="event.type", string="Event Type"),
-        'event_ok': fields.related('product_id', 'event_ok', string='event_ok', type='boolean'),
-    }
 
-    def _prepare_order_line_invoice_line(self, cr, uid, line, account_id=False, context=None):
-        res = super(sale_order_line, self)._prepare_order_line_invoice_line(cr, uid, line, account_id=account_id, context=context)
+    event_id = fields.Many2one('event.event', 'Event', help="Choose an event and it will automatically create a registration for this event.")
```
Comment (rgo-odoo) [inline addons/event_sale/models/sale_order.py]: 'string' is missing from few fields

File: addons/event_sale/models/sale_order.py [hunk 10]
```diff
@@ -1,44 +1,32 @@
...
+    event_ticket_id = fields.Many2one('event.event.ticket', 'Event Ticket', help="Choose an event ticket and it will automatically create a registration for this event ticket.")
+    event_type_id = fields.Many2one(related='product_id.event_type_id', string="Event Type")
+    event_ok = fields.Boolean(related='product_id.event_ok')
+
+    @api.model
+    def _prepare_order_line_invoice_line(self, line, account_id=False, context=None):
```
Comment (rgo-odoo) [inline addons/event_sale/models/sale_order.py]: This method seems been removed from main app and I think it is replaced by _prepare_invoice_line(...), would you please check it and override _prepare_invoice_line here.

File: addons/event_sale/models/sale_order.py [hunk 11]
```diff
@@ -1,44 +1,32 @@
...
+        res = super(SaleOrderLine, self)._prepare_order_line_invoice_line(line, account_id=account_id)
         if line.event_id:
-            event = self.pool['event.event'].read(cr, uid, line.event_id.id, ['name'], context=context)
-            res['name'] = '%s: %s' % (res.get('name', ''), event['name'])
+            res['name'] = '%s: %s' % (res.get('name', ''), line.event_id.name)
         return res
 
     @api.onchange('product_id')
```
Comment (rgo-odoo) [inline addons/event_sale/models/sale_order.py]: you may change the code of onchange to something like,

--
        values = {'event_type_id': False, 'event_ok': False}
        if self.product_id.event_ok:
            values.update(event_type_id=self.product_id.event_type_id.id,
                          event_ok=self.product_id.event_ok)
        ....

--

File: addons/event_sale/models/sale_order.py [hunk 12]
```diff
@@ -71,10 +59,21 @@ def _update_registrations(self, confirm=True, registration_data=None):
...
-                self.env['event.registration'].with_context(registration_force_draft=True).create(
+                Registration.with_context(registration_force_draft=True).create(
                     Registration._prepare_attendee_values(registration))
         return True
 
-    def onchange_event_ticket_id(self, cr, uid, ids, event_ticket_id=False, context=None):
-        price = event_ticket_id and self.pool["event.event.ticket"].browse(cr, uid, event_ticket_id, context=context).price or False
-        return {'value': {'price_unit': price}}
+    @api.multi
+    def button_confirm(self):
```
Comment (rgo-odoo) [inline addons/event_sale/models/sale_order.py]: now, this method isn't there in master.

File: addons/event_sale/views/event_views.xml [hunk 13]
```diff
@@ -1,9 +1,7 @@
 <?xml version="1.0"?>
-<openerp>
-    <data>
-
-        <record model="ir.ui.view" id="view_event_registration_ticket_search">
-            <field name="name">event.registration.ticket.search</field>
+<odoo>
+        <record id="view_event_registration_ticket_search" model="ir.ui.view">
+            <field name="name">event.registration.ticket.search.inherit</field>
```
Comment (rgo-odoo) [inline addons/event_sale/views/event_views.xml]: it should be set to 'event.registration.search.inherit'

File: addons/event_sale/views/event_views.xml [hunk 14]
```diff
@@ -43,8 +41,8 @@
             </field>
         </record>
         
-        <record model="ir.ui.view" id="view_event_form_inherit_ticket">
-            <field name="name">view_event_form_inherit_tichet</field>
+        <record id="view_event_form_inherit_ticket" model="ir.ui.view">
+            <field name="name">event.form.inherit</field>
             <field name="model">event.event</field>
             <field name="inherit_id" ref="event.view_event_form"/>
             <field name="arch" type="xml">
```
Comment (rgo-odoo) [inline addons/event_sale/views/event_views.xml]: you can remove <data> below this.

File: addons/event_sale/views/product_views.xml [hunk 15]
```diff
@@ -1,13 +1,12 @@
-<?xml version="1.0"?>
-<openerp>
-    <data>
-        <record model="ir.ui.view" id="event_sale_product_template_form">
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+        <record id="event_sale_product_template_form" model="ir.ui.view">
```
Comment (rgo-odoo) [inline addons/event_sale/views/product_views.xml]: 'name' field isn't set for this view

**Pull request merged**